restapi/src/main/java/io/stargate/web/docsapi/service/query/ExpressionParser.java

```diff
@@ -85,26 +96,99 @@ public ExpressionParser(ConditionParser predicateProvider) {
+      List<PathSegment> prependedPath,
+      Iterable<JsonNode> nodes,
+      boolean numericBooleans,
+      MutableInt nextIndex) {
     List<Expression<FilterExpression>> expressions = new ArrayList<>();
 
-    int index = 0;
-    Iterator<Map.Entry<String, JsonNode>> fields = filterJson.fields();
-    while (fields.hasNext()) {
-      Map.Entry<String, JsonNode> field = fields.next();
-      String fieldPath = field.getKey();
-      FilterPath filterPath = getFilterPath(prependedPath, fieldPath);
-      Collection<BaseCondition> fieldConditions =
-          conditionProvider.getConditions(field.getValue(), numericBooleans);
-      for (BaseCondition fieldCondition : fieldConditions) {
-        ImmutableFilterExpression expression =
-            ImmutableFilterExpression.of(filterPath, fieldCondition, index++);
-        expressions.add(expression);
+    // go through all the nodes given
+    for (JsonNode node : nodes) {
+      Iterator<Map.Entry<String, JsonNode>> fields = node.fields();
+
+      // until we have more fields left
+      while (fields.hasNext()) {
+        // get field
+        Map.Entry<String, JsonNode> next = fields.next();
+        String fieldOrOp = next.getKey();
+
+        // if $or go for Or operator
+        // if $and go for And operator
+        // otherwise go for the field resolving
+        if (Objects.equals(OR_OPERATOR, fieldOrOp)) {
+          JsonNode orChildrenNode = next.getValue();
+          Or<FilterExpression> or =
+              resolveOr(orChildrenNode, prependedPath, numericBooleans, nextIndex);
+          expressions.add(or);
+        } else if (Objects.equals(AND_OPERATOR, fieldOrOp)) {
+          JsonNode andChildrenNode = next.getValue();
+          And<FilterExpression> and =
+              resolveAnd(andChildrenNode, prependedPath, numericBooleans, nextIndex);
+          expressions.add(and);
+        } else {
+          FilterPath filterPath = getFilterPath(prependedPath, fieldOrOp);
+          Collection<BaseCondition> fieldConditions =
+              conditionProvider.getConditions(next.getValue(), numericBooleans);
+          for (BaseCondition fieldCondition : fieldConditions) {
+            ImmutableFilterExpression expression =
+                ImmutableFilterExpression.of(
+                    filterPath, fieldCondition, nextIndex.getAndIncrement());
+            expressions.add(expression);
+          }
+        }
       }
     }
 
     return expressions;
   }
 
+  private Or<FilterExpression> resolveOr(
+      JsonNode node,
+      List<PathSegment> prependedPath,
+      boolean numericBooleans,
+      MutableInt nextIndex) {
+    if (!node.isArray()) {
+      throw new ErrorCodeRuntimeException(
+          ErrorCode.DOCS_API_SEARCH_FILTER_INVALID,
+          "The $or requires an array json node as value.");
+    }
+
+    List<Expression<FilterExpression>> orConditions =
+        parse(prependedPath, node, numericBooleans, nextIndex);
```

Comment:
Now expression indexing depends on the JSON tree iteration order, which may be too obscure to the end user.

It might be worth adding explicit (but optional) selectivity hints. If a hint is _not_ provided we'd derive selectivity from relation positions. E.g. `{ field: {$eq: 1, $selectivity: 1000} }`.

For filters without an explicit selectivity, assume is it less (better) than the selectivity of all of the following filters in the same JSON array, and their nested filters.

When comparing filters with know selectivity with unknown selectivity assume that know is always better. WDYT?

I know it adds quite a bit of work to this PR, but I think we have to address the problem of how to allow users to control query plans in a simple and hopefully intuitive way since we do not have stats about real data distribution.

Reply:
I like the idea, but this is a complete other issue imo.. Can make a ticket for it.. Let's make small steps faster.. OK?

Reply:
yes. let's file an separate issue for this, but I think we'd have to address it before we enable `ORs` to be used in queries.

Reply:
https://github.com/stargate/stargate/issues/1072